scripts/gha/lint_commenter.py

```diff
@@ -0,0 +1,280 @@
+  args = parse_cmdline_args()
+  if args.repo is None:
+      args.repo=subprocess.check_output(['git', 'config', '--get', 'remote.origin.url']).decode('utf-8').rstrip('\n').lower()
+      if args.verbose:
+        print('autodetected repo: %s' % args.repo)
+  if not args.repo.startswith('https://github.com/'):
+      print('Error, only https://github.com/ repositories are allowed.')
+      exit(2)
+  (repo_owner, repo_name) = re.match(r'https://github\.com/([^/]+)/([^/.]+)', args.repo).groups()
+
+  # Get the head commit for the pull request.
+  # GET /repos/{owner}/{repo}/pulls/{pull_number}
+  request_url = 'https://api.github.com/repos/%s/%s/pulls/%s' % (repo_owner, repo_name, args.pr_number)
+  header = 'Accept: application/vnd.github.VERSION.json'
+  pr_data = json.loads(subprocess.check_output(
+      [args.curl,
+       '-s', '-X', 'GET',
+       '-H', 'Accept: application/vnd.github.v3+json',
+       '-H', 'Authorization: token %s' % args.token,
+       request_url
+      ] + ([] if not args.verbose else ['-v'])).decode('utf-8').rstrip('\n'))
+  commit_sha = pr_data['head']['sha']
+  skip_lint = False
+  if 'labels' in pr_data:
+    for label in pr_data['labels']:
+      if label['name'] == LABEL_TO_SKIP_LINT:
+        skip_lint = True
+        break
+
+  if skip_lint:
+    print('PR #%s has "%s" label, skipping lint checks' % (args.pr_number, LABEL_TO_SKIP_LINT))
+    exit(0)
+
+  if args.verbose:
+    print('Commit sha:', commit_sha)
+
+  # Get the diff for the pull request.
+  # GET /repos/{owner}/{repo}/pulls/{pull_number}
+  request_url = 'https://api.github.com/repos/%s/%s/pulls/%s' % (repo_owner, repo_name, args.pr_number)
+  header = 'Accept: application/vnd.github.VERSION.diff'
+  if args.verbose:
+    print('request_url: %s' % request_url)
+  pr_diff = subprocess.check_output(
+      [args.curl,
+       '-s', '-o', '-', '-w', '\nHTTP status %{http_code}\n',
+       '-X', 'GET',
+       '-H', header,
+       '-H', 'Authorization: token %s' % args.token,
+       request_url
+      ] + ([] if not args.verbose else ['-v'])).decode('utf-8')
+  # Parse the diff to determine the position (within the PR) of each source line.
+  # Only lint lines that refer to valid positions in the PR will be displayed.
+  # Information on what this means here:
+  # https://docs.github.com/en/rest/reference/pulls#create-a-review-comment-for-a-pull-request
+  line_reference = {}
+  file_list = []
+  pr_patch = PatchSet(pr_diff)
+  for pr_patch_file in pr_patch:
+    # Skip removed files.
+    if pr_patch_file.removed: continue
+    # Skip files that match an EXCLUDE_PATH_REGEX
+    excluded = False
+    for exclude_regex in EXCLUDE_PATH_REGEX:
+      if re.search(exclude_regex, pr_patch_file.path):
+        excluded = True
+        break
+    if excluded: continue
+    file_list.append(pr_patch_file.path)
+    pr_file_position = 1
+    line_reference[pr_patch_file.path] = {}
+    for hunk in pr_patch_file:
+      pr_file_position = pr_file_position + 1
```

Comment:
isn't this a problem that the pr_file_position starts at 2?

Reply:
Technically no, according to https://docs.github.com/en/rest/reference/pulls#create-a-review-for-a-pull-request, the @@ at the start of the diff is position 1, the next line is position 2, etc.

Reply:
Oh wait you're right! I misread it. I'll fix it.

Reply:
I decided to give up on calculating the "position" (when I tried a diff with lines removed it got way more complicated) and am now just using "line", it's much simpler. I only need to get a list of the lines touched, not track each one's position in the diff.